File: web/packages/alloy-storage/src/auth/pkce.ts
function base64url(bytes: Uint8Array): string {
  let bin = '';
  for (const b of bytes) bin += String.fromCharCode(b);
  return btoa(bin).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}

const defaultRandom = (len: number): Uint8Array => crypto.getRandomValues(new Uint8Array(len));

/** 48 random bytes → 64 base64url chars (RFC 7636 §4.1 allows 43–128). */
export function generateCodeVerifier(random: (len: number) => Uint8Array = defaultRandom): string {
  return base64url(random(48));
}

/** S256 challenge: base64url(SHA-256(verifier)). */
export async function codeChallenge(verifier: string): Promise<string> {
  const digest = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(verifier));
  return base64url(new Uint8Array(digest));
}

File: swift/Sources/AlloyStorage/Auth/PKCE.swift
import CryptoKit
import Foundation

/// PKCE helpers (RFC 7636, S256). Caseless-enum namespace, twin of auth/pkce.ts.
public enum PKCE {
  static func base64url(_ data: Data) -> String {
    data.base64EncodedString()
      .replacingOccurrences(of: "+", with: "-")
      .replacingOccurrences(of: "/", with: "_")
      .replacingOccurrences(of: "=", with: "")
  }

  /// 48 random bytes → 64 base64url chars.
  public static func generateCodeVerifier() -> String {
    var bytes = [UInt8](repeating: 0, count: 48)
    for i in bytes.indices { bytes[i] = UInt8.random(in: 0...255) }
    return base64url(Data(bytes))
  }

  /// S256 challenge: base64url(SHA-256(verifier)).
  public static func codeChallenge(_ verifier: String) -> String {
    base64url(Data(SHA256.hash(data: Data(verifier.utf8))))
  }
}
